### PR TITLE
Rework server-side entity spawning

### PR DESCRIPTION
**Pull request by tornac1234**
- [x] Fix fragment spawning
- [x] Fix nested structure spawning (e.g a bottle of water spawns inside of a supply crate inside of a wreck)
- [x] Introduce an uniform UnityEngine.Random like generation for floats (see XORRandom)
- [x] Add a button to teleport to the selected object in the Scene debugger
- [x] Implement empty objects (like lights and things which we apparently ignored in the past but therefore didn't spawn)
- [x] ~~Fix Crashes (the explosive fishes)~~ postponed to a further PR which will also fix vehicle constructors and add finitions to the entity system
- [x] Handle sub-spawners monobehaviours in placeholders groups

### CODE REVIEW

File: NitroxServer/Helper/XORRandom.cs
Comment (dartasen) [inline NitroxServer/Helper/XORRandom.cs:9]: Most of the comments can be turned into /// summary or either removed

File: NitroxServer/Helper/XORRandom.cs
Comment (dartasen) [inline NitroxServer/Helper/XORRandom.cs:13]:
```suggestion
    private const uint MT19937 = 1812433253;
```

File: NitroxServer/Helper/XORRandom.cs
Comment (dartasen) [inline NitroxServer/Helper/XORRandom.cs:26]: Don't think we're using the _ convention in Nitrox

**Review by Measurity — COMMENTED** [review]
Just a minor quick code review for now so I can look at PR 2090

File: NitroxModel/DataStructures/GameLogic/Entity.cs
Comment (tornac1234) [inline NitroxModel/DataStructures/GameLogic/Entity.cs:13]: Will reindex before merge

File: NitroxClient/GameLogic/Spawning/WorldEntities/PrefabPlaceholderEntitySpawner.cs
Comment (tornac1234) [inline NitroxClient/GameLogic/Spawning/WorldEntities/PrefabPlaceholderEntitySpawner.cs:18]: Will be completed

**Review by tornac1234 — COMMENTED** [review]
I'll wait for further reviews to push the mentionned changes

File: NitroxServer-Subnautica/GameLogic/Entities/Spawning/CrashFishBootstrapper.cs
Comment (tornac1234) [reply NitroxServer-Subnautica/GameLogic/Entities/Spawning/CrashFishBootstrapper.cs:17]: In another one

File: NitroxClient/GameLogic/Spawning/WorldEntities/SerializedWorldEntitySpawner.cs
Comment (tornac1234) [reply NitroxClient/GameLogic/Spawning/WorldEntities/SerializedWorldEntitySpawner.cs:90]: GameObjects (and other Unity components in general) can be cast to booleans but only when you explicitly do it. Here, it will either print the gameobject or a " " (space) if gameobject is null.

File: NitroxModel-Subnautica/DataStructures/GameLogic/Entities/SubnauticaUwePrefabFactory.cs
Comment (tornac1234) [reply NitroxModel-Subnautica/DataStructures/GameLogic/Entities/SubnauticaUwePrefabFactory.cs:40]: Will add those lines
```cs
// Manually went through the list of those to make this "filter"
// You can verify this by looping through all of SrcData (e.g in LootDistributionData.Initialize)
// print the prefabPath and check the TechType related to the provided classId (WorldEntityDatabase.TryGetInfo) with PDAScanner.IsFragment
```

File: NitroxModel-Subnautica/DataStructures/GameLogic/Entities/SubnauticaUwePrefabFactory.cs
Comment (tornac1234) [reply NitroxModel-Subnautica/DataStructures/GameLogic/Entities/SubnauticaUwePrefabFactory.cs:23]: Would rather make this a "TryGetPossiblePrefabs" where you return null if there are none and adapt the code using this function

File: NitroxServer/Resources/PrefabPlaceholdersGroupAsset.cs
Comment (tornac1234) [reply NitroxServer/Resources/PrefabPlaceholdersGroupAsset.cs]: We discussed in Discord that it wasn't too heavy in here so it was fine to still use struct because remaking it into a class would break the system around. Resolving the conversation

File: NitroxServer/Resources/PrefabPlaceholderAsset.cs
Comment (tornac1234) [reply NitroxServer/Resources/PrefabPlaceholderAsset.cs]: Same as for https://github.com/SubnauticaNitrox/Nitrox/pull/2093#discussion_r1435642561

**Review by Measurity — APPROVED** [review]
Minor thing: moving the `PrefabPlaceholderEntity` in `WorldPersistenceTest` to the correct `switch` (inside `WorldEntity` condition) and unit test will succeed.

**Review by dartasen — APPROVED** [review]

File: NitroxClient/GameLogic/Spawning/WorldEntities/PrefabPlaceholderEntitySpawner.cs
Comment (Jannify) [inline NitroxClient/GameLogic/Spawning/WorldEntities/PrefabPlaceholderEntitySpawner.cs:61]:
```suggestion
        Log.Error($"[{nameof(PrefabPlaceholderEntitySpawner)}] Can't find a {nameof(PrefabPlaceholdersGroup)} on parent for {entity.Id}");
```

**Review by Jannify — CHANGES_REQUESTED** [review]

File: NitroxClient/GameLogic/Spawning/WorldEntities/SerializedWorldEntitySpawner.cs
Comment (Jannify) [inline NitroxClient/GameLogic/Spawning/WorldEntities/SerializedWorldEntitySpawner.cs:28]:
```suggestion
        if (!NitroxEnvironment.IsTesting && !ProtobufSerializer.emptyGameObjectPrefab)
```

**Review by Jannify — CHANGES_REQUESTED** [review]
One test is failing still. Can be fixed with the comment in `SerializedWorldEntitySpawner.cs` and also this change in `NitroxClient.GameLogic.Spawning.WorldEntities.CrashEntitySpawner`:

```cs
private static readonly Quaternion spawnRotation = new(-0.7f, 0f, 0f, 0.7f); // Equal to Quaternion.Euler(-90f, 0f, 0f)```

**Review by dartasen — APPROVED** [review]

**Pull request merged**
